lib/helpers/versions.js

```diff
@@ -2,15 +2,84 @@ const logger = require('winston');
 const semver = require('semver');
 const stable = require('semver-stable');
 const _ = require('lodash');
+const semverUtils = require('semver-utils');
 
 module.exports = {
+  determineRangeUpgrades,
   determineUpgrades,
   isRange,
   isValidVersion,
   isFuture,
   isPastLatest,
+  computeRangedVersion,
 };
 
+function determineRangeUpgrades(dep, currentVersion, config) {
+  if (!isValidVersion(currentVersion)) {
+    logger.verbose(`${dep.name} currentVersion is invalid`);
+    return [];
+  }
+  const versions = dep.versions;
+  if (!versions || Object.keys(versions).length === 0) {
+    logger.verbose(`${dep.name} - no versions`);
+    return [];
+  }
+  const versionList = Object.keys(versions);
+  const workingVersion = semver.maxSatisfying(versionList, currentVersion);
+
+  return _.chain(versionList)
+
+    // Filter out older versions as we can't upgrade to those
+    .filter(version => semver.gt(version, workingVersion))
+
+    // Ignore unstable versions, unless the current version is unstable
+    .reject(version => config.ignoreUnstable &&
+            stable.is(workingVersion) && !stable.is(version))
+
+    // Ignore future versions, unless the current version is marked as future
+    .reject(version => config.ignoreFuture &&
+            !isFuture(versions[workingVersion]) && isFuture(versions[version]))
+
+    // Ignore versions newer than "latest", unless current version is newer than the "latest"
+    .reject(version => config.respectLatest &&
+            isPastLatest(dep, version) && !isPastLatest(dep, workingVersion))
+
+    // Process all remaining versions
+    .reduce((upgrades, newVersion) => {
```

Comment:
Okay, even though I'm really out of time to work on the minimum satisfying version feature, one idea I thought of is that you can, right before `reduce`, use Lodash's `groupBy` method like `.groupBy(semver.major)`, to group versions by major, and then, if the range is `~`, further group by `semver.minor`.

Then you could use `semver`'s `minSatisfying` method to find the lowest version in each group after constructing a pseudo version, like, for minor group `5` within major group `4`, `~4.5.0`.